pkg/querier/iterators/chunk_iterator.go

```diff
@@ -115,10 +111,10 @@ func (i *chunkIterator) AtT() int64 {
 		t, _ := i.At()
 		return t
 	case chunkenc.ValHistogram:
-		t, _ := i.AtHistogram()
+		t, _ := i.AtHistogram(i.cachedHistogram)
```

Comment:
Doesn't keep `i.cachedHistogram` like above, so no effect if it was nil. Also I'm not sure why we do this switch since `i.AtT` should be defined for all chunk iterators.

Reply:
I think that in this case we should actually cache the values that we get from `.At*`
```
func (i *chunkIterator) AtT() int64 {
	if i.cacheValid {
		return i.cachedTime
	}
	switch i.valType {
	case chunkenc.ValFloat:
		t, i.cachedValue := i.At()
		return t
	case chunkenc.ValHistogram:
		t, i.cachedHistogram := i.AtHistogram(i.cachedHistogram)
		return t
	case chunkenc.ValFloatHistogram:
		t, i.cachedFloatHistogram := i.AtFloatHistogram(i.cachedFloatHistogram)
		return t
	default:
		panic(fmt.Errorf("chunkIterator: calling AtT with unknown chunk encoding %v", i.valType))
	}
}
```

Otherwise, if we do the following sequence:
```
ts := i.AtT()
_, h := i.AtHistogram(nil)
```
this histogram will be calculated twice, and it is not necessary. WDYT?

Reply:
The last part of my last comment is actually not correct, the histogram will not be calculated twice, because `AtT()` will call `AtHistogram()`, which will internally store the cached value. 
On the other hand, when the histogram is not cached, `i.cachedHistogram` will be `nil`, so both `AtHistogram(nil)` and `AtHistogram(i.cachedValue)` will have the same effect.
Do you have any preferences @krajorama?

Reply:
Hmm, I don't think so, at least for raw chunks by the time you call At<any> the chunk was decoded, so there's no extra penalty. See:  https://github.com/prometheus/prometheus/blob/419dd265ccd198fbf312659b3671dac2ecaad510/tsdb/chunkenc/histogram.go#L917

Let's keep it simpler and do a profile. @zenador has implemented native histograms for continuous test , and also maybe we can just run some long queries and check Pyroscope.

Reply:
Let's try to switch to calling `i.AtT()` only, it *should* work.

Reply:
@krajorama do you mean something like this?
```
func (i *chunkIterator) AtT() int64 {
	if i.cacheValid {
		return i.cachedTime
	}
	return i.it.AtT()
}
```
This cannot be done because `chunk.Iterator`, i.e., the type of `i.it`, does not have the `AtT()` function.